edb/server/protocol/ai_ext.py

```diff
@@ -0,0 +1,1305 @@
+                                provider_cfg,
+                                model_name,
+                                [entry[1].decode("utf-8") for entry in part],
+                                shortening,
+                            ),
+                        )
+                        tasks[task] = part
+
+        for task, entries in tasks.items():
+            embeddings = task.result()
+            groups = itertools.groupby(entries, key=lambda e: e[2:])
+            offset = 0
+            for (rel, attr, *_), items in groups:
+                ids = [item[0] for item in items]
+                await _update_embeddings_in_db(
+                    pgconn, rel, attr, ids, embeddings, offset)
+                offset += len(ids)
+
+    return len(submit_list)
+
+
+async def _update_embeddings_in_db(
+    pgconn: pgcon.PGConnection,
+    rel: bytes,
+    attr: bytes,
+    ids: list[bytes],
+    embeddings: bytes,
+    offset: int,
+) -> int:
+    id_array = '", "'.join(uuidgen.from_bytes(ub).hex for ub in ids)
+    entries = await pgconn.sql_fetch_val(
+        f"""
+        WITH upd AS (
+            UPDATE {rel.decode()} AS target
+            SET
+                {attr.decode()} = (
+                    (embeddings.data ->> 'embedding')::edgedb.vector)
+            FROM
+                (
+                    SELECT
+                        row_number() over () AS n,
+                        j.data
+                    FROM
+                        (SELECT
+                            data
+                        FROM
+                            json_array_elements(($1::json) -> 'data') AS data
+                        OFFSET
+                            $3::text::int
+                        ) AS j
+                ) AS embeddings,
+                unnest($2::text::text[]) WITH ORDINALITY AS ids(id, n)
+            WHERE
+                embeddings."n" = ids."n"
+                AND target."id" = ids."id"::uuid
+            RETURNING
+                target."id"
+        )
+        SELECT count(*)::text FROM upd
+        """.encode(),
+        args=(
+            embeddings,
+            f'{{"{id_array}"}}'.encode(),
+            str(offset).encode(),
+        ),
+    )
+
+    return int(entries.decode())
+
+
+async def _generate_embeddings(
+    provider,
+    model_name: str,
+    inputs: list[str],
+    shortening: Optional[int],
+) -> bytes:
+    task_name = _task_name.get()
+    count = len(inputs)
+    suf = "s" if count > 1 else ""
+    logger.debug(
+        f"{task_name} generating embeddings via {model_name!r} "
+        f"of {provider.name!r} for {len(inputs)} object{suf}"
+    )
+
+    if provider.name == "builtin::openai":
+        return await _generate_openai_embeddings(
+            provider, model_name, inputs, shortening)
+    elif provider.name == "builtin::mistral":
+        return await _generate_mistral_embeddings(
+            provider, model_name, inputs, shortening)
+    else:
```

Comment:
What about Anthropic embeddings?

Reply:
Anthropic does not provide any embedding models.